### ISSUE TITLE
A few legend + color questions

### ISSUE THREAD

**Issue by ericphanson** [JuliaPlots/AlgebraOfGraphics.jl#106]
1. Is there a way to suppress the legend resulting from `style(color = ...)`?
2. Also, I've got something like `(plot1 + plot2) |> draw` and each has a `style(color = ...)`, and the colors are repeated in the legend. Anyway to get `plot1` and `plot2` to draw from the same pool of colors to avoid conflicts?
3. Or to specify the colors manually (so I can ensure they don't conflict)?

**Comment by piever** [comment]
The legend and "scales" need a little rework, so at the moment:

1. No, not really, but maybe this is less of an issue once 2. and 3. are fixed. Otherwise, I guess one could start thing of an API for that as well.
2. Mhm, in theory that should work, but I see now that maybe some of the assumptions of CategoricalArrays fail if you use them from different datasets, do you have a MWE (with a screenshot of the plot, so I can see what legend exactly you get?)
3. You can pass a whole new set of colors to use, say `visual(color = ["red", "blue", "green"])`, but not an explicit value => color mapping. OTOH, this should never really be needed if 2. is fixed I guess.

**Comment by ericphanson** [comment]
Here is an example that is pretty close to what I'm actually doing...
```julia
let
	n = 10
	cats = [rand('a':'c') for i = 1:n]
	
	table = [(idx = i, cat1 = cats[i], val = i*rand()) for _ = 1:30 for i = 1:n]
	
	s = data(table) * style(:val, :idx) * style(color = :cat1) * spec(Scatter, markersize=5, strokewidth=0)
	
	e = data((; idx=1:n, endpoint=1:n, cat2 = [string(iseven(i)) for i = 1:n])) * style(:endpoint, :idx) * style(color = :cat2) * spec(Scatter, markersize=10, strokewidth=0)
	
	(s + e) |> draw
end
```

![mwe](https://user-images.githubusercontent.com/5846501/97026525-c79a8f80-1559-11eb-8455-057989c3e6e1.png)

On 0.1.2 of AlgebraOfGraphics.

Something is very weird: if I take away the `string` call around `iseven`, I get this:

![mwe2](https://user-images.githubusercontent.com/5846501/97026642-f3b61080-1559-11eb-9007-0bc7f54d620c.png)

**Comment by ericphanson** [comment]
I'll update to master to get the fixes from #105 and see if these results change

**Comment by ericphanson** [comment]
Yep, same strange behavior on master; update code is just `style` -> `mapping` and `spec` -> `visual`:
```julia
let
	n = 10
	cats = [rand('a':'c') for i = 1:n]
	
	table = [(idx = i, cat1 = cats[i], val = i*rand()) for _ = 1:30 for i = 1:n]
	
	s = data(table) * mapping(:val, :idx) * mapping(color = :cat1) * visual(Scatter, markersize=5, strokewidth=0)
	
	e = data((; idx=1:n, endpoint=1:n, cat2 = [string(iseven(i)) for i = 1:n])) * mapping(:endpoint, :idx) * mapping(color = :cat2) * visual(Scatter, markersize=10, strokewidth=0)
	
	(s + e) |> draw
end
```

**Comment by ericphanson** [comment]
What about `layoutplot!` returning both the scene and the layout, and then `draw` just returns the `scene`? That way `|> draw` will still work, but for advanced use, you can easily get the layout (to, say, remove the legend). Not as good as an API but way more hackable in the meantime.

**Comment by piever** [comment]
> edit: `style` should be deprecated to `mapping`, not `visual`, right?

Ups, yes, fixed!

> ```julia
> let
> 	n = 10
> 	cats = [rand('a':'c') for i = 1:n]
> 	
> 	table = [(idx = i, cat1 = cats[i], val = i*rand()) for _ = 1:30 for i = 1:n]
> 	
> 	s = data(table) * mapping(:val, :idx) * mapping(color = :cat1) * visual(Scatter, markersize=5, strokewidth=0)
> 	
> 	e = data((; idx=1:n, endpoint=1:n, cat2 = [string(iseven(i)) for i = 1:n])) * mapping(:endpoint, :idx) * mapping(color = :cat2) * visual(Scatter, markersize=10, strokewidth=0)
> 	
> 	(s + e) |> draw
> end
> ```

Mhm, I'd never thought of this type of use case, so you want to somehow match colors from groupings with different variables, which take different values? Would you expect this to go on looping colors and list `false, true` in the same legend but continuing to cycle colors? That is, use 5 distinct colors in total? This could definitely be done, and it makes sense to me (compute the scales using all the data).

> Something is very weird: if I take away the string call around `iseven`, I get this...

This is intended. There is an "auto-detection" mechanism to figure out if a variable is categorical (find distinct colors) or continuous (use a colorbar). Booleans subtype numbers (for better or for worse), so you get the continuous version. `color = :myvar => categorical` would force the continuous version. This could probably be a bit more prominent in the docs.

> Also, if I try passing explict colors, they all end up the second one

That is somewhat intended, to pass it to both you should do something like

`(s + e) * visual(color = [:red, :green, :blue])`

> What about `layoutplot!` returning both the scene and the layout, and then `draw` just returns the `scene`? That way `|> draw` will still work, but for advanced use, you can easily get the layout (to, say, remove the legend). Not as good as an API but way more hackable in the meantime.

Yes, that makes sense, to have `layoutplot` to return both. Your are not the first one to ask for this, and I guess it makes sense to have a high-level interface and then return a tweakable object.

**Comment by ericphanson** [comment]
> Mhm, I'd never thought of this type of use case, so you want to somehow match colors from groupings with different variables, which take different values? 

Yeah, essentially I've got two categories, and I want to color the data by one of them, and I also happen to want to indicate the end of the data, which I was using black dots for. But then I wanted to also show the second category somehow, and since layouting wasn't working (fixed by #105), I decided to color the end-marker by the second category.

> Would you expect this to go on looping colors and list false, true in the same legend but continuing to cycle colors? That is, use 5 distinct colors in total? This could definitely be done, and it makes sense to me (compute the scales using all the data).

Yeah, that's what I wanted to happen, five distinct colors. I actually wanted to also remove the first category from the legend (since there's a lot of them, and it matters more that you can tell groups apart rather than specifically which group is which). Now I'm using layouting for the first category and deleting the legend by hand, so I don't have a legend at all.

> This is intended. There is an "auto-detection" mechanism to figure out if a variable is categorical (find distinct colors) or continuous (use a colorbar). Booleans subtype numbers (for better or for worse), so you get the continuous version. color = :myvar => categorical would force the continuous version. This could probably be a bit more prominent in the docs.

Ah ok, good to know! I did see the `=> categorical` bits in the tutorial, but when I omitted it (on my string categories) nothing changed, so I kind of forgot about it.

> That is somewhat intended, to pass it to both you should do something like `(s + e) * visual(color = [:red, :green, :blue])`

Ah ok, makes sense.

**Comment by ericphanson** [comment]
Closing this since I don't think there's anything concrete to do here :). Thanks very much for the quick answers @piever!
